core/src/sumeragi/main_loop.rs

```diff
@@ -40,31 +40,15 @@ pub struct Sumeragi {
     pub block_time: Duration,
     /// The maximum number of transactions in the block
     pub max_txs_in_block: usize,
-    /// [`TransactionValidator`] instance that we use
-    pub transaction_validator: TransactionValidator,
     /// Kura instance used for IO
     pub kura: Arc<Kura>,
     /// [`iroha_p2p::Network`] actor address
     pub network: IrohaNetwork,
     /// Receiver channel.
-    // TODO: Mutex shouldn't be required and must be removed
-    pub message_receiver: Mutex<mpsc::Receiver<MessagePacket>>,
+    pub message_receiver: mpsc::Receiver<MessagePacket>,
     /// Only used in testing. Causes the genesis peer to withhold blocks when it
     /// is the proxy tail.
     pub debug_force_soft_fork: bool,
-}
-
-impl Debug for Sumeragi {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Sumeragi")
-            .field("public_key", &self.key_pair.public_key())
-            .field("peer_id", &self.peer_id)
-            .finish()
-    }
-}
-
-/// Internal structure that retains the state.
-pub struct State {
```

Comment:
@appetrosyan much of the state inside like `previous_block_hash`, `current_block_hash` and so on essentially can be removed, because we already have this information inside `wsv` or `finalized_wsv`.

It's something i want to do as part of different PR if you don't mind.

Additionally since now we move `Sumeragi` into `run` function it's should be possible to do state machines as you supposed.